File: 0743-network-delay-time/0743-network-delay-time.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        List<List<Pair<Integer, Integer>>> adj = new ArrayList<>(n+1);
        for(int i = 0; i <= n; i++) adj.add(new ArrayList<>());
        for(int[] time : times){
            adj.get(time[0]).add(new Pair(time[1], time[2]));
        }
        Queue<Integer> q = new LinkedList<>();
        q.offer(k);
        int[] minTime = new int[n+1];
        Arrays.fill(minTime, Integer.MAX_VALUE);
        minTime[k] = 0;
        while(!q.isEmpty()){
            int curNode = q.remove();
            int stepTime = minTime[curNode];

            for(Pair<Integer, Integer> nNode : adj.get(curNode)){
                if(stepTime+nNode.getValue() < minTime[nNode.getKey()]){
                    minTime[nNode.getKey()] = stepTime+nNode.getValue();
                    q.offer(nNode.getKey());
                }
            }
        }

        int time = 0;
        for(int i = 1; i <= n; i++){
            if(minTime[i] == Integer.MAX_VALUE)return -1;
            time = Math.max(time, minTime[i]);
        }
        return time;
    }
}

File: 0743-network-delay-time/0743-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj = [[] for _ in range(n + 1)]

        for time in times:
            adj[time[0]].append((time[1], time[2]))

        q = deque([k])
        min_time = [float('inf')] * (n + 1)
        min_time[k] = 0

        while q:
            cur_node = q.popleft()
            step_time = min_time[cur_node]

            for n_node in adj[cur_node]:
                if step_time + n_node[1] < min_time[n_node[0]]:
                    min_time[n_node[0]] = step_time + n_node[1]
                    q.append(n_node[0])

        time = 0
        for i in range(1, n + 1):
            if min_time[i] == float('inf'):
                return -1
            time = max(time, min_time[i])

        return time